cpp/src/join/distinct_hash_join.cu

```diff
@@ -0,0 +1,386 @@
+
+  auto constexpr tile_size   = HashTable::cg_size;
+  auto constexpr window_size = HashTable::window_size;
+
+  auto idx          = cudf::detail::grid_1d::global_thread_id() / tile_size;
+  auto const stride = cudf::detail::grid_1d::grid_stride() / tile_size;
+  auto const block  = cg::this_thread_block();
+
+  // CG-based probing algorithm
+  if constexpr (tile_size != 1) {
+    auto const tile = cg::tiled_partition<tile_size>(block);
+
+    auto constexpr flushing_tile_size = cudf::detail::warp_size / window_size;
+    // random choice to tune
+    auto constexpr flushing_buffer_size = 2 * flushing_tile_size;
+    auto constexpr num_flushing_tiles   = DISTINCT_JOIN_BLOCK_SIZE / flushing_tile_size;
+    auto constexpr max_matches          = flushing_tile_size / tile_size;
+
+    auto const flushing_tile    = cg::tiled_partition<flushing_tile_size>(block);
+    auto const flushing_tile_id = block.thread_rank() / flushing_tile_size;
+
+    __shared__ cuco::pair<cudf::size_type, cudf::size_type>
+      flushing_tile_buffer[num_flushing_tiles][flushing_tile_size];
+    // per flushing-tile counter to track number of filled elements
+    __shared__ cudf::size_type flushing_counter[num_flushing_tiles];
+
+    if (flushing_tile.thread_rank() == 0) { flushing_counter[flushing_tile_id] = 0; }
+    flushing_tile.sync();  // sync still needed since cg.any doesn't imply a memory barrier
+
+    while (flushing_tile.any(idx < n)) {
+      bool active_flag = idx < n;
+      auto const active_flushing_tile =
+        cg::binary_partition<flushing_tile_size>(flushing_tile, active_flag);
+      if (active_flag) {
+        auto const found = hash_table.find(tile, *(iter + idx));
+        if (tile.thread_rank() == 0 and found != hash_table.end()) {
+          auto const offset = atomicAdd_block(&flushing_counter[flushing_tile_id], 1);
+          flushing_tile_buffer[flushing_tile_id][offset] = cuco::pair{
+            static_cast<cudf::size_type>(found->second), static_cast<cudf::size_type>(idx)};
+        }
+      }
+      flushing_tile.sync();
+      if (flushing_counter[flushing_tile_id] + max_matches > flushing_buffer_size) {
+        flush_buffer(flushing_tile,
+                     flushing_counter[flushing_tile_id],
+                     flushing_tile_buffer[flushing_tile_id],
+                     counter,
+                     build_indices,
+                     probe_indices);
+        flushing_tile.sync();
+        if (flushing_tile.thread_rank() == 0) { flushing_counter[flushing_tile_id] = 0; }
+        flushing_tile.sync();
+      }
+
+      idx += stride;
+    }  // while
+
+    flushing_tile.sync();
+    if (flushing_counter[flushing_tile_id] > 0) {
+      flush_buffer(flushing_tile,
+                   flushing_counter[flushing_tile_id],
+                   flushing_tile_buffer[flushing_tile_id],
+                   counter,
+                   build_indices,
+                   probe_indices);
+    }
+  }
+  // Scalar probing
```

Comment:
It'd be helpful to clarify that this branch is related to the cooperative group size. I didn't realize that originally (hence my earlier comment). It's pretty far away from the `if constexpr` so I had lost mental context by this point.
```suggestion
  // Scalar probing for CG size 1
```